fix: Make TxidFromString() respect string_view length

hodlinator opened the pull request:
### Problem

Prior to this, `TxidFromString()` was passing `string_view::data()` into `uint256S()` which meant it would only receive the a naked `char*` pointer and potentially scan past the `string_view::length()` until it found a null terminator (or some other non-hex character).

Appears to have been a fully dormant bug as callers were either passing a string literal or `std::string` directly to `TxidFromFromString()`, meaning a null terminator always existed at `pointer[length()]`. Bug existed since original merge of `TxidFromString()`.

### Solution

Make `uint256S()` (and `base_blob::SetHex()`) take and operate on `std::string_view` instead of `const char*` and have `TxidFromString()` pass that in.


(PR was prompted by comment in https://github.com/bitcoin/bitcoin/pull/30377#issuecomment-2208857200 (referring to https://github.com/bitcoin/bitcoin/pull/28922#discussion_r1404437378)).

maflcko reviewed (commented):
left some comments

hodlinator committed:
fix: Make TxidFromString() respect string_view length

Prior to this, passing string_view::data() into uint256S() meant the latter would only receive the a naked char* pointer and potentially scan past the string_view::length() until it found a null terminator (or some other non-hex character).

Appears to have been a fully dormant bug as callers were either passing a string literal or std::string directly to TxidFromFromString(), meaning null terminator always existed at pointer[length()]. Bug existed since original merge of TxidFromString(), discussed in https://github.com/bitcoin/bitcoin/pull/28922#discussion_r1404437378.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
-    // skip 0x
-    if (psz[0] == '0' && ToLower(psz[1]) == 'x')
-        psz += 2;
+    str = util::LeftTrimStringView(str);
```
Thanks for the feedback, I really enjoyed removing the overloads based on the other comments! :)

Messed up the first force-push somehow, here is a comparison between what you reviewed and the current state:
https://github.com/bitcoin/bitcoin/compare/be23937392195c773811fdfc0d723783b2dace67..fcec222ae9386342ff1342e096255ca1ff74964d

> Also, could add a unit test that truncated input is parsed?

Do you mean that `base_blob::SetHex()` allows parsing "half bytes" like `"0xF"`, while `ParseHex()` would fail on `"F"` and require `"0F"` (maybe `"F0"`)?

hodlinator commented on src/rest.cpp:
```diff
@@ -789,8 +789,9 @@ static bool rest_getutxos(const std::any& context, HTTPRequest* req, const std::
-            std::string strOutput = uriParts[i].substr(uriParts[i].find('-')+1);
+            const std::string_view part{uriParts[i]};
+            const std::string_view strTxid{part.substr(0, part.find('-'))};
+            const std::string_view strOutput{part.substr(strTxid.size()+1)};
```
Hm.. may have managed to figure out why it would not fail before. Old code:
```C++
std::string strTxid = uriParts[i].substr(0, uriParts[i].find('-'));
std::string strOutput = uriParts[i].substr(uriParts[i].find('-')+1);
```
If `.find('-')` returned `npos` and `npos = size_type(-1)`, then `+1` would make it `0`, so it would not throw in `substr()`. This would happen for an empty `uriParts[i]`, or non-empty not containing `-` (in this latter case it might have failed parsing `strOutput` if encountering `A-F`).

hodlinator replied:
Improved testing and error reporting in 01fa92928e6f9509935840a8c57d29774a0d14a1.

Made 3d0aec126f8103310b741a5c21e94cf537fcd191 use `util::Split`. It allocates a vector for each parsed UTXO but is more readable than my former approach.

hodlinator commented on src/uint256.h:
```diff
@@ -112,21 +111,11 @@ class uint256 : public base_blob<256> {
- * in dangerously catching uint256(0).
+/* uint256 from std::string_view.
+ * This is a separate function because the constructor uint256(std::string_view str) can result
+ * in dangerously catching uint256(0) via std::string_view(const char*).
```
e9a436b7796cb285730d179fb3317ad50234ab04 changed it to: `This is not a uint256 constructor because of historical fears of uint256(0) resolving to a NULL string and crashing.`

hodlinator commented on src/rest.cpp:
```diff
@@ -789,8 +789,11 @@ static bool rest_getutxos(const std::any& context, HTTPRequest* req, const std::
+            const auto strOutput{utxo_pair[1]};
 
             if (!ParseInt32(strOutput, &nOutput) || !IsHex(strTxid))
                 return RESTERR(req, HTTP_BAD_REQUEST, "Parse error: couldn't read \"<hexadecimal txid>-<decimal output index>\" from \"" + uriParts[i] + "\"");
```
Updated to avoid changing existing error strings. Also reordered commits, let me know if you prefer I break off the last 2 into a separate PR.

hodlinator replied:
Dropped 2 commits from this PR.

hodlinator commented on src/util/string.h:
```diff
@@ -68,11 +68,11 @@ std::vector<T> Split(const Span<const char>& sp, char sep)
 [[nodiscard]] inline std::string_view TrimStringView(std::string_view str, std::string_view pattern = " \f\n\r\t\v")
 {
-    std::string::size_type front = str.find_first_not_of(pattern);
+    std::string_view::size_type front = str.find_first_not_of(pattern);
     if (front == std::string::npos) {
```
Seems very harmless to correct the type. But yeah, should have fixed the `npos` if we go with this change.

I'll remove it from this PR for now if it goes outside of the "PR style" of the project.

hodlinator commented on src/util/string.h:
```diff
@@ -83,8 +83,8 @@ std::vector<T> Split(const Span<const char>& sp, char sep)
-    if (str.substr(0, prefix.size()) == prefix) {
-        return str.substr(prefix.size());
+    if (str.starts_with(prefix)) {
+        str.remove_prefix(prefix.size());
```
`base_blob` now calls `RemovePrefixView()` which it did not previously do. I want to make it as efficient as possible compared to the prior `base_blob` implementation, hence this change.

maflcko reviewed (approved):
ACK c3a9c71c7077324bf0aa40f834f7537a14619340 🐞

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK c3a9c71c7077324bf0aa40f834f7537a14619340 🐞
0oe/WjDO2mSwVc5Blutf21rEvpYjnqCdQdO76x2UhqHObkUZghAnaWmbpcKtyptM6F8bw8dIl8IU3+TJAkcnBg==
```

</details>

stickies-v commented on src/uint256.h:
```diff
@@ -112,21 +111,11 @@ class uint256 : public base_blob<256> {
- * This is a separate function because the constructor uint256(const char*) can result
- * in dangerously catching uint256(0).
+/* uint256 from std::string_view.
+ * This is not a uint256 constructor because of historical fears of uint256(0)
```
nit: I interpret "historical fears" as "the fears are no longer applicable", in which case I think it should just be a constructor and this helper function removed. But, it seems like until C++23 this is indeed an issue (even if the ambiguous overload with the `uint8_t` should prevent compilation), so the current wording is confusing to me.

Suggested alternative:

```
/* uint256 from std::string_view.
 * This is not a uint256 constructor to avoid the ambiguity of
 * uint256(0) being interpreted as initialization from a nullptr.
 *
 * This helper can be replaced with a constructor when we require c++23,
 * which prohibits constructing a string_view from a nullptr, see
 * https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2166r1.html
 */
```

hodlinator replied:
Thank you for the correction! Applied with some minor adjustments.

hodlinator replied:
Reset back to prior version, see https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1682274787.

stickies-v reviewed (commented):
Approach ACK c3a9c71c7077324bf0aa40f834f7537a14619340

hodlinator commented on src/uint256.h:
```diff
@@ -112,21 +111,15 @@ class uint256 : public base_blob<256> {
+ * This is not a uint256 constructor to avoid the ambiguity of uint256(0) being
+ * interpreted as initialization from a nullptr.
+ *
+ * This helper can be replaced with a constructor when we require C++23, which
```
Tried:
```diff
diff --git a/src/test/uint256_tests.cpp b/src/test/uint256_tests.cpp
index 2053a42d24..e8dfb03d18 100644
--- a/src/test/uint256_tests.cpp
+++ b/src/test/uint256_tests.cpp
@@ -119,6 +119,10 @@ BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
     BOOST_CHECK(uint160(R1S) == R1S);
     BOOST_CHECK(uint160(ZeroS) == ZeroS);
     BOOST_CHECK(uint160(OneS) == OneS);
+
+    uint256{0};
+    uint256{NULL};
+    uint256{nullptr};
 }
 
 BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
diff --git a/src/uint256.h b/src/uint256.h
index 19faaca91f..39b776ed0f 100644
--- a/src/uint256.h
+++ b/src/uint256.h
@@ -107,6 +107,7 @@ public:
     constexpr uint256() = default;
     constexpr explicit uint256(uint8_t v) : base_blob<256>(v) {}
     constexpr explicit uint256(Span<const unsigned char> vch) : base_blob<256>(vch) {}
+    explicit uint256(std::string_view str) { SetHex(str); }
     static const uint256 ZERO;
     static const uint256 ONE;
 };
```

G++13.2.0 warns against both NULL and nullptr. `0` and `NULL` resolve to the `uint256(uint8_t v)` ctor:
```
test/uint256_tests.cpp: In member function ‘void uint256_tests::basics::test_method()’:
test/uint256_tests.cpp:124:17: warning: passing NULL to non-pointer argument 1 of ‘constexpr uint256::uint256(uint8_t)’ [-Wconversion-null]
  124 |     uint256{NULL};
      |                 ^
...
./uint256.h:108:40: note:   declared here
  108 |     constexpr explicit uint256(uint8_t v) : base_blob<256>(v) {}
      |                                ~~~~~~~~^
test/uint256_tests.cpp:125:20: warning: argument 1 null where non-null expected [-Wnonnull]
  125 |     uint256{nullptr};
      |                    ^
...
/nix/store/14c6s4xzhy14i2b05s00rjns2j93gzz4-gcc-13.2.0/include/c++/13.2.0/string_view:140:7: note: in a call to function ‘constexpr std::basic_string_view<_CharT, _Traits>::basic_string_view(const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>]’ declared ‘nonnull’
  140 |       basic_string_view(const _CharT* __str) noexcept
      |       ^~~~~~~~~~~~~~~~~
...
/nix/store/14c6s4xzhy14i2b05s00rjns2j93gzz4-gcc-13.2.0/include/c++/13.2.0/bits/char_traits.h:409:32: warning: argument 1 null where non-null expected [-Wnonnull]
  409 |         return __builtin_strlen(__s);
      |                ~~~~~~~~~~~~~~~~^~~~~
/nix/store/14c6s4xzhy14i2b05s00rjns2j93gzz4-gcc-13.2.0/include/c++/13.2.0/bits/char_traits.h:409:32: note: in a call to built-in function ‘long unsigned int __builtin_strlen(const char*)’
```
Clang++(17.0.6) warns as well:
```
test/uint256_tests.cpp:124:13: warning: implicit conversion of NULL constant to 'uint8_t' (aka 'unsigned char') [-Wnull-conversion]
  124 |     uint256{NULL};
      |     ~~~~~~~ ^~~~
      |             '\0'
test/uint256_tests.cpp:125:13: warning: null passed to a callee that requires a non-null argument [-Wnonnull]
  125 |     uint256{nullptr};
      |             ^~~~~~~
2 warnings generated.
```
Both continue compilation and segfault when running the tests.

Does CI have a stricter warnings-as-errors policy?

stickies-v replied:
Thanks for your comprehensive input here @maflcko. You're right that this is safe as long as one CI job fails, there's no need to ensure this fails on every single compiler, which indeed makes the fear historical. My apologies for the back and forth.

Your suggestion for a `FromHex()` method also seems better than a `std::string_view` ctor.

hodlinator replied:
I agree `uint256::FromHex()` is more self-explanatory than `uint256S()`. Making it return `std::optional` might be good if we want stricter parsing and non-fatal failures. This PR feels large enough for now though.

ryanofsky reviewed (approved):
Code review ACK c3a9c71c7077324bf0aa40f834f7537a14619340

I think the changes are all good, but the PR and the most important commit "fix: Make TxidFromString() respect string length" (c3a9c71c7077324bf0aa40f834f7537a14619340) are lacking a good description that actually says what the bug is.

I think they should say that currently `TxidFromString` takes a string_view parameter, but internally it is calling the `uint256S` function which expects a nul-terminated string. If `TxidFromString` is called with a string that not contain `\0`, the implementation will read the string and potentially uninitialized memory after it until it finds any character which is not a hex digit.

ryanofsky commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
     size_t digits = 0;
-    while (::HexDigit(psz[digits]) != -1)
-        digits++;
+    for (const char c : str) {
+        if (::HexDigit(c) == -1)
+            break;
+        ++digits;
+    }
```
Thanks, I did not know this function will be removed. It would be helpful to mention in the PR description that it will be removed, so reviewers can know how worried to be about the quality of the code.

stickies-v reviewed (approved):
ACK c3a9c71c7077324bf0aa40f834f7537a14619340,

> are lacking a good description that actually says what the bug is.

I agree, I spent quite a bit of time getting to the bottom of this and more documentation would have sped that up

> but internally it is calling the uint256S function which expects a nul-terminated string

Maybe I'm being too pedantic (or wrong), but I'm not sure this is 100% correct? I don't think the `uint256S(std::string_view str)` `str` needs to be null-terminated. Rather, I think the problem is that by removing the string length data in `TxidFromstring` (by passing `std::string_view::data` instead of `std::string_view`), `uint256S` has to construct a `std::string_view` from the `const char*` without length information, which is dangerous (as you describe) when the initial `string_view` was not null-terminated, e.g. because it is a substring?

For example, this test (with a non-null terminated string) fails before c3a9c71c7077324bf0aa40f834f7537a14619340 and succeeds after:

```cp
//txid "0x74d681e0e03bafa802c8aa084379aa98d9fcd632ddc2ed9782b586ec87451f20"
const char arr[66] = {
    '7', '4', 'd', '6', '8', '1', 'e', '0', 'e', '0', '3', 'b', 'a', 'f', 'a', '8',
    '0', '2', 'c', '8', 'a', 'a', '0', '8', '4', '3', '7', '9', 'a', 'a', '9', '8',
    'd', '9', 'f', 'c', 'd', '6', '3', '2', 'd', 'd', 'c', '2', 'e', 'd', '9', '7',
    '8', '2', 'b', '5', '8', '6', 'e', 'c', '8', '7', '4', '5', '1', 'f', '2', '0',
    'a', 'b' // extra 2 characters
};
const char* ptr = arr;  // not null-terminated
std::string_view view{ptr, 64};
BOOST_CHECK_EQUAL(TxidFromString(view).ToString(), view);
```

ryanofsky commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
     size_t digits = 0;
-    while (::HexDigit(psz[digits]) != -1)
-        digits++;
+    for (const char c : str) {
+        if (::HexDigit(c) == -1)
+            break;
+        ++digits;
+    }
```
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1682892102

> review note: it seems the `while` loop was replaced with `for` loop purely for readability, I don't see any functional change here.

To be sure, replacing the previous while loop is an important part of the fix. If the previous while loop was kept, it could iterate past the end of the of `string_view` and read uninitialized memory. This for loop (and the while loop I suggested in https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683086730) avoid the bug by stopping iteration at the end of the `string_view`.

stickies-v replied:
> it could iterate past the end of the of `string_view` and read uninitialized memory.

Right, because on master we rely on stopping when `HexDigit()` returns a `-1`, which is fine when the string is null-terminated but is not safe otherwise. Thanks for pointing this out, this also clarifies my misunderstanding about your [earlier comment](https://github.com/bitcoin/bitcoin/pull/30436/#pullrequestreview-2186195988).

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/30436#pullrequestreview-2185910679

> > but internally it is calling the uint256S function which expects a nul-terminated string
> 
> Maybe I'm being too pedantic (or wrong), but I'm not sure this is 100% correct? I don't think the `uint256S(std::string_view str)` `str` needs to be null-terminated.

Yes, both explanations are saying the same thing but mine is unnecessarily confusing. The `uint256S` function "expects" a nul-terminated string because it accepts a bare `char*`  argument, and in C bare `char*` strings are almost always nul-terminated. But it does not "need" a nul-terminated string. The string can be terminated with any character other than `0123456789AaBbCcDdEeFf`, as mentioned in the next sentence.

stickies-v commented:
Side note: we can use [bugprone-suspicious-stringview-data-usage](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/suspicious-stringview-data-usage.html) (from unreleased clang-19) to catch this going forward, when we have clang-19. I tried running clang-tidy-19 locally from master branch for longer than I probably should have, and eventually gave up.

As a poor man's alternative, I looked for other stringview data misusage based on some rough grepping, and couldn't find anything, but that was definitely not an exhaustive search either.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
-    // skip 0x
-    if (psz[0] == '0' && ToLower(psz[1]) == 'x')
-        psz += 2;
+    str = util::TrimStringView(str);
```
> Note: This line isn't directly equivalent to previous code because it trims whitespace from the end of the string, not just the beginning, but the net effect is the same because any whitespace at the end of the string would be ignored anyway.

@ryanofsky yes, that's why my first pushed implementation added `LeftTrimStringView` in 8f4293d892358ce145eb9a099f2e5f256c4d9c73. But got pushback in https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1675658574 and trimming the end of the string means we save checking one character later when range-for iterating over the trimmed `string_view` to count digits.

stickies-v commented:
re-ACK 788fe9cc9ab979c5e14f544ae05dc46436892b81 

No changes except for commit message update.

l0rinc reviewed (changes requested):
Thanks for tackling these, please my comments, most of them originally posted to https://github.com/bitcoin/bitcoin/pull/30482/files

hodlinator commented:
> Also, you can adjust the pull request description with a motivation and fix. Otherwise, it is just two random links, and reviewers will have a hard time seeing the motivation and gist of the changes.

Done. Thank you for you patience @maflcko.

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -102,6 +102,8 @@ BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
     BOOST_CHECK(uint256S(R1L.ToString()) == R1L);
     BOOST_CHECK(uint256S("   0x"+R1L.ToString()+"   ") == R1L);
+    BOOST_CHECK(uint256S("   0x"+R1L.ToString()+"-trash;%^&   ") == R1L);
+    BOOST_CHECK(uint256S("\t \n  \n \f\n\r\t\v\t   0x"+R1L.ToString()+"  \t \n  \n \f\n\r\t\v\t ") == R1L);
```
Would fully agree if we were introducing a new test function, but as it is in the middle of existing `BOOST_CHECK`s, I think it's fine. Won't sweat it if I don't show up in the `git blame` after the `BOOST_CHECK_EQUAL`-refactor. And I'm happy to review such a follow-up PR.

(Original change here was suggested in https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1675658574)

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
 
 template <unsigned int BITS>
-void base_blob<BITS>::SetHex(const char* psz)
+void base_blob<BITS>::SetHex(std::string_view str)
```
Not sure about `const`ing `string_view` parameters, could only find one pre-existing case of that pattern (`GetBuriedDeployment()`). Can sympathize with forcing introduction of new names over mutation. I like how you named the first digit `hi` (maybe `high` is more precise).

Ultimately, my modification of the function is meant to be a straightforward transformation of the original code. Since it has received a few ACKs already, I am going to put off changing it for now. But might draw some inspiration from your version if I need to change it for other reasons.

paplorinc replied:
We can get a reack, let's not focus on that.
I'm NACK 788fe9cc9ab979c5e14f544ae05dc46436892b81 -ing because of https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686336262 anyway.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686328446

I don't really think it makes sense to make declare string views const, since the data they point to is already immutable, and the argument is passed by value so no changes to the view inside the function could affect the caller. I think the const just adds verbosity. Not 100% sure about this and haven't thought about it much, but would want to know a specific reason for using const here, or see some broader recommendation somewhere that would explain why to do this.

paplorinc replied:
> the data they point to is already immutable

The string part, but not the leading and trailing offsets.


The `const` signals the intention of the `string_view`.
By declaring it as `const std::string_view str` I'm announcing that its value will be constant throughout the method, right?
But if it's just `std::string_view str`, I can call `str.remove_prefix` or `str.remove_prefix`, which modifies the internals of the view, so after the call its internal state is different.

hodlinator replied:
@paplorinc I see `hi`/`lo` is consistent with `secp256k1_uint128`, so I retract my `high` thoughts.

It strikes me that the `hi` part of your suggestion is not the one being shifted `<< 4`. Maybe it should be called `lo`?

paplorinc replied:
you're absolutely right, edited the example!

paplorinc replied:
Let me try a different angle: if a method is inline, I would expect to be able to copy-paste its content to the call site directly (at least conceptually, to understand its effect).
If the method is pure in a functional sense (including that it doesn't even mutate its parameters), it's a lot easier to reason about its overall behavior.
Does this resonate more?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1688616750

> Let me try a different angle: if a method is inline, I would expect to be able to copy-paste its content to the call site directly (at least conceptually, to understand its effect). If the method is pure in a functional sense (including that it doesn't even mutate its parameters), it's a lot easier to reason about its overall behavior. Does this resonate more?

I actually learned something new about this looking at bc3139e5a5f1bf81e4bb53417157f0f2cef86716 from #30377: apparently our CI tools do not complain if you declare parameters `const` in the function definition but omit the `const` in the function declarations. I previously thought this would cause warnings or lint errors if definitions and declarations weren't exactly the same. So it seems like this provides a good compromise. You can avoid confusing callers by not including a `const` in the declaration that doesn't mean or do anything, while keeping const in the definition if you want to make it obvious that a parameter is not mutated.

I still think declaring const for function parameters passed by value doesn't make sense as a general rule, but it can make sense in certain cases for the reasons you explained.

I also learned there is a clang-tidy rule [`readability-avoid-const-params-in-decls`](https://clang.llvm.org/extra/clang-tidy/checks/readability/avoid-const-params-in-decls.html) that might be nice for us to enable.

paplorinc replied:
> function parameters passed by value doesn't make sense as a general rule

[Google cpp guide](https://google.github.io/styleguide/cppguide.html#Use_of_const ) mostly aligns with your view on this. However, I approach this from a more functional programming perspective, where I want to infer the possible state changes a method can perform based solely on its signature - or at least what it cannot do.

Thanks for following up on this, @ryanofsky!

ryanofsky replied:
> I approach this from a more functional programming perspective, where I want to infer the possible state changes a method can perform based solely on its signature - or at least what it cannot do.

Yes, that's definitely good and desirable, but my point is that when you add `const` to a c++ function parameter that is passed by value (not passed by reference) the `const` **does not become part of the function signature**. It is completely ignored by callers and has no effect on what inputs can be passed to the function or what changes the function can make to those inputs (because c++ functions already cannot modify inputs which are passed by value).

The only effect of adding top-level const on function parameters passed by value is to make local argument variables inside the function definition constant instead of mutable. This can have benefits as well as drawbacks, just like functional programming sometimes has benefits over imperative programming, and sometimes has drawbacks. An important drawback that can happen when variables are immutable is it can lead to proliferation of similarly named variables containing slightly different values because it is not possible to replace existing values, leading to confusion between variables and bugs like https://github.com/bitcoin/bitcoin/pull/30386#discussion_r1683742656. I just think it is good idea to look at things on a case by case basis and not assume functional approaches are always or usually better.

l0rinc replied:
We're in 100% agreement here, thanks for detailing your view.

In this particular case my thinking was:
> initial code, parameter (copy of the original) is mutated locally, we have to keep track of state changes in our head to understand the method's behavior:
```C++
std::string_view RemovePrefixView2(std::string_view str, std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        str.remove_prefix(prefix.size());
    }
    return str;
}
```
> adding const makes parameter mutations forbidden - we have to rethink the algo (i.e. code smell):
```C++
std::string_view RemovePrefixView3(const std::string_view str, const std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        str.remove_prefix(prefix.size()); //  error: 'this' argument to member function 'remove_prefix' has type 'const std::string_view' (aka 'const basic_string_view<char>'), but function is not marked const
    }
    return str;
}
```
> we can return a copy of the view, knowing that `str` didn't change state throughout the method
```C++
std::string_view RemovePrefixView4(const std::string_view str, const std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        return str.substr(prefix.size());
    }
    return str;
}
```
which enables:
```C++
std::string_view RemovePrefixView5(const std::string_view str, const std::string_view prefix)
{
    return str.starts_with(prefix)
         ? str.substr(prefix.size())
         : str;
}
```
At which point we might as well remove the `const` from the params, since the new impl is now trivial:
```C++
std::string_view RemovePrefixView5(std::string_view str, std::string_view prefix)
{
    return str.starts_with(prefix) ? str.substr(prefix.size()) : str;
}
```

ryanofsky replied:
Makes sense that const can expose things even in simple cases and lead to insights and improvements like the one you are suggesting. In this case, I think "if starts with prefix call remove_prefix" might be more straightforward than "if starts with prefix call substr and return portion of the string after the prefix" but the difference is not great and I take your point in general const can have benefits in places like this.

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -115,6 +117,8 @@ BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
     BOOST_CHECK(uint160S("0x"+MaxS.ToString()) == MaxS);
     BOOST_CHECK(uint160S(R1S.ToString()) == R1S);
     BOOST_CHECK(uint160S("   0x"+R1S.ToString()+"   ") == R1S);
+    BOOST_CHECK(uint160S("   0x"+R1S.ToString()+"-trash;%^&   ") == R1S);
```
I did, of course, I'm more interested here in your opinion.

paplorinc replied:
Since we're modifying the function I'm suggesting more coverage for the cases, based on what other hex parsers were testing for. It's not a difficult or unreasonable request.

paplorinc replied:
> But no need to hold up a bugfix based on a test-only minor nit.

agree, this isn't the reason for my NACK.

paplorinc replied:
If you think it's redundant, please mark this as resolved.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
     size_t digits = 0;
-    while (::HexDigit(psz[digits]) != -1)
-        digits++;
+    for (const char c : str) {
+        if (::HexDigit(c) == -1)
+            break;
+        ++digits;
+    }
```
It is random access, but it's also contiguous, allowing for the safe range-`for`. If I had to do it again I might have lifted the `break;` up to the same line as the `if`. If I need to retouch and others support that change I might do that to make it slightly more terse like yours.

paplorinc replied:
You can of course redo it, I didn't review your change for us to ponder on what could have been...

ryanofsky replied:
> You can of course redo it, I didn't review your change for us to ponder on what could have been...

I think the while loop is a slightly better but range-for is also ok.

I still do think as long as you are going to update the loop to respect the length of the input (instead of blowing past it) it would be better update the loop to respect the length of the output as well, and not pointlessly read more data than can actually be used, as suggested https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683086730.

Not very important though, especially if this function will be deleted later.

hodlinator replied:
> I still do think as long as you are going to update the loop to respect the length of the input (instead of blowing past it) it would be better update the loop to respect the length of the output as well, and not pointlessly read more data than can actually be used, as suggested https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683086730.

I tried making the loop counting `digits` actually respect the output length (`WIDTH*2` here) as well.
```C++
    for (const char c : trimmed) {
        if (digits == WIDTH*2 || ::HexDigit(c) == -1) break;
        ++digits;
    }
```
This uncovered some weird behavior. Suddenly this check started failing:
```C++
    BOOST_CHECK_EQUAL(R1S, uint160S(R1ArrayHex));
```
Because of how the following loop in `SetHex()` parses digits from the end, it turns out that without the limit on `digits`, **if we were given more hex digits than fit, we would read the rightmost ones rather than the leftmost ones.** :exploding_head:

(The behavior of reading the rightmost hex digits predates this PR, and I'd rather not go down that rabbit hole, as the function will hopefully be removed - https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683226283).

paplorinc commented:
NACK 788fe9cc9ab979c5e14f544ae05dc46436892b81, because of https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686336262 mostly.

hodlinator commented on src/util/string.h:
```diff
@@ -83,8 +83,8 @@ std::vector<T> Split(const Span<const char>& sp, char sep)
-    if (str.substr(0, prefix.size()) == prefix) {
-        return str.substr(prefix.size());
+    if (str.starts_with(prefix)) {
+        str.remove_prefix(prefix.size());
```
* `std::string_view::remove_prefix()` itself invites mutation.
* The names `TrimStringView()` and `RemovePrefixView()` imply mutation.
* As I said in https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686494537, I could only find one other case of `const` `string_view` parameters in the entire codebase. I'm sorry if you are used to another way. It may well be a better way, but I'd rather not change more pre-existing things like `TrimStringView()` and `RemovePrefixView()` taking non-const `string_view` copies.

I do not think introducing test cases to prove that the C++ compiler is properly copying a given parameter provides much value.

paplorinc replied:
> std::string_view::remove_prefix() itself invites mutation.

no, we din't mutate before, we returned a new view. We're also doing that now, but the copy is done implicitly in the parameters. The original `substr` way made more sense - like we're doing with `TrimStringView` as well.

hodlinator replied:
The reason for using the standard C++ `remove_prefix()` function inside of the pre-existing `RemovePrefixView()` over `substr()` was covered here:
https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1679138796 
> `base_blob` now calls `RemovePrefixView()` which it did not previously do. I want to make it as efficient as possible compared to the prior `base_blob` implementation, hence this change.

To be even more clear than I was that time - please compare the implementations of [substr()](https://github.com/gcc-mirror/gcc/blob/61715e9340ab8941d40d62158fe437e9dbe3e068/libstdc%2B%2B-v3/include/std/string_view#L334) and [remove_prefix()](https://github.com/gcc-mirror/gcc/blob/61715e9340ab8941d40d62158fe437e9dbe3e068/libstdc%2B%2B-v3/include/std/string_view#L295) in GCC.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30436/files#r1686336262

> I find it quite confusing that we're sneaking in a mutation of the copy

paplorinc, maybe it would help if you could say more specifically what you find confusing about this code? At least to me the new implementation of this function seems more straightforward than the previous implementation, though I don't think there is a big difference.

I don't love the name of the `RemovePrefixView` function, because that name sounds like it actually does modify the string view like `remove_prefix` does, instead of just returning a new string view with the prefix stripped. I think a better name for this function would just be `StripPrefix`. But that's also a subjective opinion, and renaming would be beyond the scope of this PR.

paplorinc replied:
I'm not sure what you're saying, both are the same, one creates a copy via the params (unless it's inlined, I guess, that's why I find it confusing) and modifies the copied instance, the other returns a new copied instance.
I really love speed optimizations that also make the code more readable, but this does neither.

given:
```C++
static void SetHexBenchmark(benchmark::Bench& bench)
{
    uint256 hash;
    bench.run([&] {
        hash.SetHex("1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef");
        hash.SetHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef");
        ankerl::nanobench::doNotOptimizeAway(hash);
    });
}
```

with 788fe9cc9ab979c5e14f544ae05dc46436892b81 we're getting `2940379.88 ops/` on average of 2 runs:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.85 |        2,933,826.25 |    0.6% |      0.01 | `SetHexBenchmark`

and

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              339.34 |        2,946,933.51 |    0.1% |      0.01 | `SetHexBenchmark`



with `return str.substr(prefix.size());` it's `2936915.645 ops/s` on average for 2 runs:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.99 |        2,932,626.59 |    0.6% |      0.01 | `SetHexBenchmark`

and

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.00 |        2,941,204.70 |    0.0% |      0.01 | `SetHexBenchmark`


I.e. the difference of the averages is a lot smaller than the variation of individual runs.

which is basically the same speed as `master`:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              352.95 |        2,833,265.72 |    0.3% |      0.01 | `SetHexBenchmark`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              356.66 |        2,803,757.30 |    1.4% |      0.01 | `SetHexBenchmark`

------

let's focus on making the code cleaner and more maintainable - if it's also faster, great, but let's not make it harder to reason about an maintain for an imaginary speed gain.

paplorinc replied:
> it would help if you could say more specifically what you find confusing about this code

I'll try to rephrase.
Firstly, as you've mentioned as well, the name of the method hints at mutation, but it actually returns a copy of the view.
Second, it's inline + pass by value mutation, which implies it's not a simple search&replace type inline, since the parameters aren't consts, but it also can't just mutate the original values passed as parameters, so the copy of the parameter is implicit. Making the parameters const and returning a new view directly solves most of these, i.e. you can just copy the internals of the method and paste it to its usage (which helps in theory to understand how the method will behave).
Third, this seems like a premature and untested optimization, which should likely be done in a separate PR, in case the code isn't more readable.

paplorinc replied:
> This pull request has more than a 100 comments for less lines of code changed

I did all the changes I requested during review (even provided diffs to speed it up), they're not complicated.

paplorinc replied:
> No, from reading the function signature, one knows that a copy is being made when the argument is put in as a function parameter

but it's an inline function, there is no function parameter.

hodlinator replied:
Dropped `RemovePrefixView()` modification.

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -58,13 +58,7 @@ static std::string ArrayToString(const unsigned char A[], unsigned int width)
-    return rv;
-}
-inline uint160 uint160S(const std::string& str)
+inline uint160 uint160S(std::string_view str)
```
See https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686533487

paplorinc replied:
I'm not sure why the type is relevant here, what I'm saying is that we had a [const](https://learn.microsoft.com/en-us/cpp/cpp/const-cpp) guarantee before which we can keep.

paplorinc replied:
it's not very important in this particular method, but `inline` + `const` could help with the actual inlining - it's also clearer to the reader to understand what happens when mutating the parameter, since after inlining it should ideally avoid the copying (which is easy to do for a const).
At least this was my thinking for recommending it, let me know if I'm mistaken.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
-void base_blob<BITS>::SetHex(const char* psz)
+void base_blob<BITS>::SetHex(std::string_view str)
 {
     std::fill(m_data.begin(), m_data.end(), 0);
```
Good find. Will do if I need to retouch that commit for other reasons.

paplorinc replied:
My other suggestions make this irrelevant, we can zero during iteration as well

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686345800

Calling SetNull, or skipping zeroing as suggested seem like improvements, but no strong opinion. Also think the other suggestions to use begin / end and simplify the loop are good, so added :+1: 

I can also understand other point of view of wanting to change code as little as possible while fixing a bug, and wanting to avoid asking people to rereview code that is changing and potentially introduce new bugs. But whatever you want to do here seems fine.

paplorinc replied:
if we're afraid to modify code, it's usually a code smell, which points to not enough test coverage.
Instead of being afraid, we should improve our code coverage.
We can't take a tank to battle if we're afraid we'll scratch it.

ryanofsky reviewed (approved):
Code review ACK 788fe9cc9ab979c5e14f544ae05dc46436892b81. No changes since my last review other than a commit message change.

This might be ready for merge since it has 3 acks and is definitely an improvement over the status quo. But I haven't read through the comments yet or looked into the concerns behind the one nack. It also does seem like this PR might still be a little rough around the edges, so maybe it wouldn't hurt to iterate a little more.

re: https://github.com/bitcoin/bitcoin/pull/30436#issue-2405108982

> Prior to this, `TxidFromString()` was passing `string_view::data()` into `uint256S()` which meant it would only receive the a naked `char*` pointer and potentially scan past the `string_view::length()` until it found a null terminator.

This is a helpful summary, but not 100% accurate. `uint256S()` does currently require a terminator, but it doesn't have to be a `\0` terminator. Any character other than `01234567890ABCDEFabcdef` will work as a terminator and stop reading the string.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
     unsigned char* pend = p1 + WIDTH;
     while (digits > 0 && p1 < pend) {
-        *p1 = ::HexDigit(psz[--digits]);
+        *p1 = ::HexDigit(str[--digits]);
```
I like your version slightly more, will do if I need to re-touch the commit and others agree.

Sorry for being difficult. Wish your review had come in sooner.

paplorinc replied:
> Wish your review had come in sooner

nothing is lost yet, we're all on the same side, people will glady reack :)

ryanofsky reviewed (approved):
Responded to a bunch of comments, but after reading them I do think this PR looks ok in its current form, and is an improvement.

Would be happy to merge as is, or re-ack if there are more changes.

hodlinator commented:
> This is a helpful summary, but not 100% accurate. `uint256S()` does currently require a terminator, but it doesn't have to be a `\0` terminator. Any character other than `01234567890ABCDEFabcdef` will work as a terminator and stop reading the string.

(Updated summary slightly).

hodlinator commented:
1b7bfbab04f73ca6b9256b4e01f1eec3da5cb541 - Applied many of the suggestions from @paplorinc and @ryanofsky.
Hopefully this will achieve even more ACKs.

Added commit replacing `BOOST_CHECK()` with `BOOST_CHECK_*()` versions in **uint256_tests.cpp**.

Dropped change to `RemovePrefixView()` as @paplorinc's benchmark shows no improvement https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686678529

Added `TxidFromString()`-test, demonstrating fix as per @ryanofsky's suggestion.

Took @paplorinc's version of the final loop in `SetHex()`, but changed it to range-for:
```C++
    for (auto& elem : m_data) {
        auto hi = (digits > 0) ? ::HexDigit(trimmed[--digits]) : 0;
        elem = (digits > 0) ? hi | (::HexDigit(trimmed[--digits]) << 4) : hi;
    }
```

Replaced first loop in `SetHex()` with `find_if()`:
```C++
    size_t digits = std::ranges::find_if(trimmed, [](char c) { return ::HexDigit(c) == -1; }) - trimmed.begin();
```
Let me know if you prefer:
```C++
    size_t digits = 0;
    for (const char c : trimmed) {
        if (::HexDigit(c) == -1) break;
        ++digits;
    }
```

Documented endianness of hex strings as per @paplorinc's suggestion + added test.

Sorry about the extra rebase, I'm holding one of the git commands wrongly. [Comparison between previously reviewed and current](https://github.com/bitcoin/bitcoin/compare/788fe9cc9ab979c5e14f544ae05dc46436892b81..1b7bfbab04f73ca6b9256b4e01f1eec3da5cb541)

maflcko reviewed (changes requested):
weak Nack, based on https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1687659582 . Not sure why a reviewed pull request for a fix is changed to add in refactoring that isn't needed for the fix and can be  done later, if it isn't redundant at all.

Please keep "refactor" changes to the absolute minimum in real code.

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,19 @@ std::string base_blob<BITS>::GetHex() const
-            p1++;
-        }
+    const auto trimmed = util::RemovePrefixView(util::TrimStringView(str), "0x");
+    // Note: if we are passed a greater number of digits than would fit as bytes
+    // in m_data, we will be discarding the leftmost ones.
+    // str="12bc" in a WIDTH=1 m_data => m_data[] == "\0xbc", not "0x12".
+    size_t digits = std::ranges::find_if(trimmed, [](char c) { return ::HexDigit(c) == -1; }) - trimmed.begin();
```
> Not sure about changing code that will be obsolete in a few commits. It seems wasteful to waste review on code that will be changed or deleted down the line. It seems better to touch every line of code only once (or the least number of times possible), than to repeatedly refactor it, and then remove the "feature".

The function is not being removed as of #30482.

> For example, by removing the `std::fill(m_data.begin(), m_data.end(), 0);`, two `SetHex` calls now may result in a different value, because the previous value is not zeroed, but leaked into the new value.

This is not true, the final loop iterates over the whole of `m_data`, filling zeroes on the remaining elements when running out of `digits`. It felt much more robust to me using a range-for, and cleaner then the previous `p1`/`pend`.

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -73,54 +73,54 @@ inline uint160 uint160S(const std::string& str)
 BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
 {
-    BOOST_CHECK(1 == 0+1);
+    BOOST_CHECK_EQUAL(1, 0+1);
```
I wonder what this does exactly :)

hodlinator replied:
Indeed.. saw it too. Seems like temporary code that somehow made it into the test. Guess I could remove if I retouch.

paplorinc commented on src/uint256.cpp:
```diff
@@ -18,39 +18,19 @@ std::string base_blob<BITS>::GetHex() const
+    // Note: if we are passed a greater number of digits than would fit as bytes
+    // in m_data, we will be discarding the leftmost ones.
+    // str="12bc" in a WIDTH=1 m_data => m_data[] == "\0xbc", not "0x12".
+    size_t digits = std::ranges::find_if(trimmed, [](char c) { return ::HexDigit(c) == -1; }) - trimmed.begin();
```
+1

paplorinc commented on src/test/transaction_tests.cpp:
```diff
@@ -1030,10 +1030,10 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
+    // end of the substring.
     BOOST_CHECK_EQUAL(TxidFromString(std::string_view("ABCD1234").substr(0, 4)).ToString(),
-        "00000000000000000000000000000000000000000000000000000000abcd1234");
+        "000000000000000000000000000000000000000000000000000000000000abcd");
```
+1, verified, fails as expected in previous commit
nit, could probably be simplifed to:
```suggestion
    BOOST_CHECK_EQUAL(TxidFromString(std::string_view("ABCD1234", 4)),
                      uint256S("000000000000000000000000000000000000000000000000000000000000abcd"));
```

hodlinator replied:
Removed the `.substr()` now but kept the raw string instead of wrapping the second parameter in `uint256S()`.

l0rinc reviewed (approved):
Thanks for fixing this!
ACK bfcc016825d8febf366a6b5eec5ed193c2c313bc

ryanofsky committed:
test: Add test for TxidFromString() behavior

hodlinator committed:
refactor: Make uint256_tests no longer use deprecated BOOST_CHECK()

hodlinator committed:
test: uint256 - Garbage suffixes and zero padding

hodlinator committed:
refactor: Change base_blob::SetHex() to take std::string_view

Clarify that hex strings are parsed as little-endian.

hodlinator commented:
Credit @ryanofsky as author of `TxidFromString()` test with a minor transformation.

Use `BOOST_CHECK_EQUAL_COLLECTIONS()`.

`SetHex()` - Largely reset back to previously reviewed implementation with 3 ACKs. Sacrificing both my `find_if()` and @paplorinc & mine's very nice final loop. Keeping the `trimmed` `string_view` variable and `const`ed the parameter as a hat-tip to @paplorinc.

As I was changing `SetHex()` I initially forgot to bring back `std::fill(m_data.begin(), m_data.end(), 0);`, **but unit tests still succeeded**, so I added a test to catch that.

Hope this strikes a good balance. Thank you again for your patience and hopefully my intuition in doing async review online has been refined.

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -129,140 +127,147 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
-    TmpL.SetHex(R2L.ToString());   BOOST_CHECK(TmpL == R2L);
-    TmpL.SetHex(ZeroL.ToString()); BOOST_CHECK(TmpL == uint256());
+    BOOST_CHECK_EQUAL(TmpL, R1L);
+    // Verify previous values don't persist when setting to truncated string.
```
+1

paplorinc commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
     size_t digits = 0;
-    while (::HexDigit(psz[digits]) != -1)
-        digits++;
+    for (const char c : trimmed) {
+        if (::HexDigit(c) == -1) break;
+        ++digits;
+    }
```
nit: was
```suggestion
 while (digits < trimmed.size() && ::HexDigit(trimmed[digits]) != -1)
        ++digits;
```
also controversial?

hodlinator replied:
Agree the diff would have been simpler, in this instance I was attempting to revert back to what already had been discussed and ACK:ed before you joined the review. Thanks for helping improve this PR!

l0rinc reviewed (approved):
ACK 09ce3501fa2ea2885a857e380eddb74605f7038c

Note: adding co-authorship for the commits that reviewers have contributed to is a good way to make sure all participants are motivated to continue doing it.

ryanofsky reviewed (approved):
Code review ACK 09ce3501fa2ea2885a857e380eddb74605f7038c. I think the current code changes are about as small as you could make to fix the bug without introducing a string copy, and the surrounding test improvements are all very nice and welcome.

It seems like may be some suggestions from paplorinc that could be responded to, but they are pretty minor. So this looks ready for merge.

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -129,140 +128,140 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
-    BOOST_CHECK(TmpL.begin() + 32 == TmpL.end());
-    BOOST_CHECK(GetSerializeSize(R1L) == 32);
-    BOOST_CHECK(GetSerializeSize(ZeroL) == 32);
+    BOOST_CHECK_EQUAL_COLLECTIONS(R1L.begin(), R1L.end(), R1Array, R1Array + R1L.size());
+    BOOST_CHECK_EQUAL_COLLECTIONS(TmpL.begin(), TmpL.end(), R1Array, R1Array + TmpL.size());
+    BOOST_CHECK_EQUAL_COLLECTIONS(R2L.begin(), R2L.end(), R2Array, R2Array + R2L.size());
+    BOOST_CHECK_EQUAL_COLLECTIONS(ZeroL.begin(), ZeroL.end(), ZeroArray, ZeroArray + ZeroL.size());
+    BOOST_CHECK_EQUAL_COLLECTIONS(OneL.begin(), OneL.end(), OneArray, OneArray + OneL.size());
```
I originally did that for the +32 case, but wasn't working for the +20

hodlinator replied:
Agree it is bad form to use variable from the left side of the comparison in the right side.

Tested `std::end(R1Array)` but it would be off by 1 thanks to the null-terminator in the string, resulting in a size/distance-mismatch between the collections/iterators being tested.

Follow-up fixing this issue: #30526

stickies-v commented on src/test/util/setup_common.cpp:
```diff
@@ -79,6 +79,18 @@ const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 
+std::ostream& operator<<(std::ostream& os, const arith_uint256& num)
+{
+    os << ArithToUint256(num).ToString();
```
Is the `ArithToUint256` conversion necessary?

hodlinator replied:
Good point, I missed the `base_uint::ToString()` method. Not as familiar with `arith_uint256` as I am with `uint256` yet.

Should probably also document endianness of `arith_uint256` strings to mirror the added comments to `uint256` in this PR.

hodlinator replied:
Follow-up fixing this issue: #30526

stickies-v reviewed (commented):
post-merge ACK 09ce3501fa2ea2885a857e380eddb74605f7038c

hodlinator commented on src/uint256.cpp:
```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
+        *p1 = ::HexDigit(str[--digits]);
         if (digits > 0) {
-            *p1 |= ((unsigned char)::HexDigit(psz[--digits]) << 4);
+            *p1 |= ((unsigned char)::HexDigit(str[--digits]) << 4);
```
Endianness was documented in a re-touch - too bad I got it wrong the first time. :sweat_smile: Follow-up in #30526.
